### ISSUE TITLE
Documentation for `processExited()` conflicts with the implementation

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3521]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)| @zseil reported|
|-|-|
|Trac ID|trac#3521|
|Type|defect|
|Created|2008-11-02 11:36:45Z|
|Branch|https://github.com/twisted/twisted/tree/processexited-failure-3521|

Documentation for `twisted.internet.interfaces.IProcessProtocol.processExited()`,
`twisted.internet.protocol.ProcessProtocol.processExited()` and the process howto
all state that the `reason` parameter to `processExited()` should be a
`twisted.python.Failure` instance, but in reality it is always either a
`twisted.internet.error.ProcessDone` or a `twisted.internet.error.ProcessTerminated`
instance (see `twisted.internet._baseprocess.BaseProcess._callProcessExited()`).
The tests don't detect this because they all use deferreds, which
automatically convert all exceptions to Failures.

<details><summary>Searchable metadata</summary>

```
trac-id__3521 3521
type__defect defect
reporter__zseil zseil
priority__normal normal
milestone__Twisted_8_2 Twisted-8.2
branch__branches_processexited_failure_3521 branches/processexited-failure-3521
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1225625805000000 1225625805000000
changetime__1226809798000000 1226809798000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [25274]) Branching to 'processexited-failure-3521'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

Tests improved, implementation repaired, build results available http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/processexited-failure-3521

This is an incompatible change to an API introduced after 8.1.  Fixing it before 8.2 would be ideal.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_4"></a>|@radix set owner to @exarkun|
|-|-|

I *may* have instead just put a 'self.assertTrue(isinstance(x, Failure))' in the implementation of the test's processExited right before the .errback call, since that would test the new behavior while requiring significantly less (somewhat confusing) changes to the rest of the test code.

But, that's no biggy. If you want to do that, go ahead, otherwise feel free to merge as it is.

Those build results look a bit weird. Do you know why ubuntu-py2.4-select "failed" 0 tests? Perhaps more relevant, are the process-related failures in ubuntu64-py2.5-select related to this change?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

I wanted to avoid raising an exception in `processExited` since the reactor is above that code on the stack and will handle and log any exception, preventing the test from finishing except due to timeout.

I did some new builds.  All the problems are gone (windows has the `twisted.test.test_process.ProcessTestCase.testManyProcesses` errors now though, but they're pre-existing; freebsd has the fcntl EAGAIN problem, also pre-existing).  They looked like intermittent things.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [25366]) Merge processexited-failure-3521

Author: exarkun
Reviewer: radix
Fixes: [#3521](https://github.com/twisted/twisted/issues/3521)

Fix the implementation of `IReactorProcess` so that it passes a `Failure` to
`IProcessProtocol.processExited` instead of passing an exception.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
